src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -563,13 +566,16 @@ private BoundLambda ReallyInferReturnType(NamedTypeSymbol delegateType, Immutabl
                 diagnostics: diagnostics);
             Binder lambdaBodyBinder = new ExecutableCodeBinder(_unboundLambda.Syntax, lambdaSymbol, ParameterBinder(lambdaSymbol, binder));
             var block = BindLambdaBody(lambdaSymbol, lambdaBodyBinder, diagnostics);
-
-            var result = new BoundLambda(_unboundLambda.Syntax, block, diagnostics.ToReadOnlyAndFree(), lambdaBodyBinder, delegateType, inferReturnType: true)
+            var returnTypes = ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)>.GetInstance();
+            BoundLambda.BlockReturns.GetReturnTypes(returnTypes, block);
+            var inferredReturnType = BoundLambda.InferReturnType(returnTypes, lambdaBodyBinder.Compilation, lambdaBodyBinder.Conversions, delegateType, lambdaSymbol.IsAsync);
+            returnTypes.Free();
+            var result = new BoundLambda(_unboundLambda.Syntax, _unboundLambda, block, diagnostics.ToReadOnlyAndFree(), lambdaBodyBinder, delegateType, inferredReturnType)
             { WasCompilerGenerated = _unboundLambda.WasCompilerGenerated };
 
-            HashSet<DiagnosticInfo> useSiteDiagnostics = null; // TODO: figure out if this should be somehow merged into BoundLambda.Diagnostics.
-            var returnType = result.InferredReturnType(ref useSiteDiagnostics) ?? LambdaSymbol.InferenceFailureReturnType;
-            lambdaSymbol.SetInferredReturnType(result.RefKind, returnType);
+            // TODO: Should InferredReturnType.UseSiteDiagnostics be merged into BoundLambda.Diagnostics?
```

Comment:
>TODO [](start = 15, length = 4)

PROTOTYPE

Reply:
This comment is a rewording of the existing `TODO` comment above. It is not related to this change.

---
In reply to: [193189704](https://github.com/dotnet/roslyn/pull/27283#discussion_r193189704) [](ancestors = 193189704)